Label merging + MoE

ryspark opened the pull request:
Allow finetuning a single task-type head for multiple datasets, e.g. one unified decoder head for vessel detection and marine infrastructure detection.
* Supports label merging, i.e. if task 1 has N outputs and task 2 has M outputs, then they are merged and offset appropriately, with the labels being concatenated into a N+M dimensional target
* Models are saved in this label-merged format, see `rslearn_projects` for scripts to unmerge saved models
* Supports task embedding layers to distinguish between different tasks, including a simple per-task learnable embedding lookup and a multi-headed cross-attention layer along the spatial dimensions (Q = per-task learnable embedding + fixed spatial positional encodings, K/V = encoder output – similar to the attention evals in pretraining `helios`)
* **NOTE:** Though sometimes detect/segment tasks have redundant `unknown` or `background` classes, those are kept separate when merging labels between tasks, since an `unknown` from one task might correspond to a known label from another task. Hopefully, this won't make much of a difference due to the task embeddings.

Other new changes (sorry for the long PR):
* Adds support for `DecoderTrunk`, shared weights between the encoder and task-specific decoder heads. This module can inject arbitrary auxiliary losses during training.
* Adds [soft Mixture-of-Experts](https://arxiv.org/abs/2308.00951) as well as infrastructure for extending to other types of decoder trunks.
* For validation with `load_all_patches=False`, the random patch selection per instance is now consistent across different runs (`rng` set based on epoch number).
* **BREAKING:** Changes the signature of `SingleTaskModel` and `MultiTaskModel` to output a dictionary with keys `outputs`, `loss_dict`, and any other keys, instead of a tuple of `(outputs, loss_dict)`. This is useful for passing arbitrary information from the decoders to `RslearnLightningModule`. I also added hooks `on_train_forward`, `on_val_forward`, and `on_test_forward` in `RslearnLightningModule` (by default no-ops) that allow users to access and manipulate this output dictionary (e.g. for custom logging with model weights passed from the decoders, etc).

favyen2 commented on rslearn/models/multitask.py:
```diff
@@ -32,34 +35,50 @@ def __init__(
         """
         super().__init__()
         self.lazy_decode = lazy_decode
+        self.decoder_to_target: dict[str, list[str]] = decoder_to_target  # type: ignore
```
`self.decoder_to_target` could be None here since `decoder_to_target` can be None. I don't think the `type: ignore` should be used here.

favyen2 commented on rslearn/train/tasks/multi_task.py:
```diff
@@ -143,24 +181,84 @@ def __init__(self, task_name: str, metric: Metric):
+            pred: the prediction
+            target: the target
+        """
+        offset = self.task_label_offsets[self.task_name]["offset"]
```
What if self.task_label_offsets[self.task_name] is not set? Seems like it should default to 0 or avoid calling separate_task_labels.

ryspark replied:
Will update this, though `task_label_offsets` is a key generated by a script in `rslp` so adheres to a pretty rigid format in practice

favyen2 reviewed (commented):
If you have one task with 3 classes and one with 4 classes, I can see how you can use the task embedding to change what gets predicted, but if you want to output 7 probabilities and use the task_label_offsets too then how would you do that? I guess I couldn't figure out how you might increase the number of classes for that decoder since it seems to come from one of the tasks.

ryspark commented:
> Also, would it make sense (and is it possible) to handle the offsets also within the torch.nn.Module instead of in the Task?

Yeah, this would probably make sense, the main requirements are that inputs/targets are properly merged/unmerged prior to model forward pass and prior to metric computation. I could move this to `MultiTaskModel` or a copy, I didn't originally because I wanted to keep all data operations in `tasks/`. But, if you think this is cleaner, I can refactor it.

> If you have one task with 3 classes and one with 4 classes, I can see how you can use the task embedding to change what gets predicted, but if you want to output 7 probabilities and use the task_label_offsets too then how would you do that? 

What do you mean by outputting 7 probabilities while using `task_label_offsets`– like if we wanted 3 + 4 + 7 total outputs across 3 tasks? This is also probably confusing without the context from `rslp`, but configs using `task_label_offsets` are never really expected to be handwritten, it's just a way to pass information from the individual configs to a merged decoder one. It's a little clunky, but merging the configs programmatically seemed much better than trying to re-initialize decoders on the go.

ryspark commented:
Updated since the last review:
1) Moved all label merging functionality to `MultiTaskMergedModel` (away from the `MultiTask`), subclassed from `MultiTaskModel`.
2) Added MoE implementation (mostly copied from `https://github.com/lucidrains/soft-moe-pytorch`, see `rslearn/models/moe` directory).
3) Improved `DecoderTrunk` support for post-encoder MoE, task embeddings, etc. (this might make more sense in the encoder at some point, but for now I left it in the decoder so it's not tied to any specific backbone).
4) **BREAKING**: changed the signature of SingleTaskModel and MultiTaskModel to output a dictionary with keys outputs, loss_dict, and any other keys, instead of a tuple of (outputs, loss_dict).

favyen2 commented on rslearn/train/lightning_module.py:
```diff
@@ -345,7 +356,7 @@ def predict_step(
         """
         inputs, _, _ = batch
-        outputs, _ = self(inputs)
+        outputs, *_ = self(inputs)
```
I think it outputs dict now.

favyen2 commented on rslearn/models/multitask.py:
```diff
@@ -85,44 +150,242 @@ def apply_decoder(
+
+
+class MultiTaskMergedModel(MultiTaskModel):
+    """MultiTaskMerged model.
```
Instead of repeating the docstring it should focus on the difference from MultiTaskModel (merging part) with some detail about what it means to merge and why the user might want to do that.
